File: kuhn_cfr/trainer.py
from constants import *
from random import shuffle
from node import Node

class Trainer:
    def __init__(self):
        self.node_map = {}

    def train(self, *, iterations):
        cards = [1, 2, 3]
        utility = 0
        for _ in range(iterations):
            shuffle(cards)
            utility += self.cfr(cards, EMPTY_HISTORY, 1, 1)
        print(f"Average game value: {utility/iterations}")
        for node in self.node_map.values():
            print(node)

    def cfr(self, cards: list, history: str, cf_reach_probability_0: float, cf_reach_probability_1: float):
        plays = len(history)
        player = plays % 2
        opponent = 1 - player

        # Terminal states
        if plays > 1:
            terminal_pass = history[-1] == 'p'
            double_bet = history[-2:] == "bb"
            player_card_higher = cards[player] > cards[opponent]
            if terminal_pass:
                if history == "pp":
                    return 1 if player_card_higher else -1
                else:
                    return 1
            elif double_bet:
                return 2 if player_card_higher else -2

        information_set = str(cards[player]) + history
        node = self.node_map.get(information_set)
        if node is None:
            node = Node(information_set)
            self.node_map[information_set] = node
        
        strategy = node.get_strategy()
        realization_weigth = cf_reach_probability_0 if player == 0 else cf_reach_probability_1
        node.update_average_strategy(strategy, realization_weigth)
        
        action_utilities = [0 for _ in range(NUM_ACTIONS)]
        node_utility = 0
        for a in range(NUM_ACTIONS):
            next_history = history + ACTIONS[a]
            if player == 0:
                action_utilities[a] = -self.cfr(cards, next_history, cf_reach_probability_0 * strategy[a], cf_reach_probability_1)
            else:
                action_utilities[a] = -self.cfr(cards, next_history, cf_reach_probability_0, cf_reach_probability_1 * strategy[a])
            node_utility += action_utilities[a] * strategy[a]
        
        for a in range(NUM_ACTIONS):
            regret = action_utilities[a] - node_utility
            weigthed_regret = regret * cf_reach_probability_1 if player == 0 else regret * cf_reach_probability_0
            node.accumulate_action_regret(a, weigthed_regret)
        
        return node_utility

        

File: kuhn_cfr_cpp/trainer.cpp
#include "trainer.h"
#include "time.h"
#include "constants.h"
#include <iostream>
#include <string>

void Trainer::Train(int p_iterations)
{
    int cards[] = {1, 2, 3};
    double utility = 0;
    srand(time(nullptr));
    for (size_t i = 0; i < p_iterations; i++)
    {
        // Shuffling
        for (int c1 = (sizeof(cards) / sizeof(int)) - 1; c1 > 0; c1--)
        {
            int c2 = rand() % (c1 + 1);
            int tmp = cards[c1];
            cards[c1] = cards[c2];
            cards[c2] = tmp;
        }

        double startingCfProbabilities[] = {1, 1};
        utility += cfr(cards, EMPTY_HISTORY, startingCfProbabilities);
    }
    std::cout << "Average game value:" << utility / p_iterations << std::endl;
    for (std::map<std::string, Node *>::iterator it = m_nodeMap.begin(); it != m_nodeMap.end(); it++)
    {
        std::cout << (it->second)->ToString() << std::endl;
    }
}
double Trainer::cfr(int p_cards[], const std::string &p_history, double p_cfReachProbability[2])
{
    int plays = p_history.length();
    int player = plays % 2;
    int opponent = 1 - player;

    // Terminal states
    if (plays > 1)
    {
        bool terminalPass = p_history.back() == 'p';
        bool doubleBet = p_history.substr(p_history.size() - 2) == "bb";
        bool playerCardHigher = p_cards[player] > p_cards[opponent];
        if (terminalPass)
        {
            if (p_history == "pp")
            {
                return playerCardHigher ? 1 : -1;
            }
            else
            {
                return 1;
            }
        }
        else if (doubleBet)
        {
            return playerCardHigher ? 2 : -2;
        }
    }

    std::string informationSet = std::to_string(p_cards[player]) + p_history;
    std::map<std::string, Node *>::iterator it = m_nodeMap.find(informationSet);
    Node *node = nullptr;
    if (it != m_nodeMap.end())
    {
        node = m_nodeMap[informationSet];
    }
    else
    {
        node = new Node(informationSet);
        m_nodeMap[informationSet] = node;
    }
    double *strategy = node->GetStrategy();
    node->UpdateCumulativeStrategy(strategy, p_cfReachProbability[player]);

    double actionUtilities[] = {0, 0};
    double nodeUtility = 0;
    for (size_t a = 0; a < NUM_ACTIONS; a++)
    {
        std::string nextHistory = p_history + ACTIONS[a];
        double nextCfReachProbability[] = {p_cfReachProbability[0], p_cfReachProbability[1]};
        if (player == 0)
        {
            nextCfReachProbability[0] *= strategy[a];
        }
        else
        {
            nextCfReachProbability[1] *= strategy[a];
        }
        actionUtilities[a] = -cfr(p_cards, nextHistory, nextCfReachProbability);
        nodeUtility += actionUtilities[a] * strategy[a];
    }
    for (size_t a = 0; a < NUM_ACTIONS; a++)
    {
        double regret = actionUtilities[a] - nodeUtility;
        double weigthedRegret = regret * p_cfReachProbability[1 - player];
        node->AccumulateActionRegrets(a, weigthedRegret);
    }
    delete[] strategy;
    return nodeUtility;
}
